Add auth with certificates (#609)
hashicorp/web-unified-docs: content/well-architected-framework/docs/docs/secure-systems/secure-applications/authenticate-certificates/manage-tls-iac.mdx

---
page_title: Manage TLS with Infrastructure as Code
description: Manage TLS certificates using Infrastructure as Code (IaC) tools to automate deployment and renewal processes.
---

# Manage TLS with Infrastructure as Code

> If your use case requires TLS certificates, but does not require signing by established certificate authorities, you can provision self-signed certificates. A self-signed certificate is a certificate that is issued by the same entity that requests it, and does not have the same chain of trust as a certificate signed by an authority. These certificates are typically used in non-production cases, like development and testing. One way to provision these kinds of certificates is through infrastructure as code. This allows consistent and repeatable certificate provisioning without the overhead of involving established CAs like Vault for signing and verifying. 

Reviewer: ```suggestion
If you need Transport Layer Security (TLS) certificates but do not need signing by established certificate authorities, you can provision self-signed certificates. A self-signed certificate is a certificate that is issued by the same entity that requests it and does not have the same chain of trust as a certificate signed by an authority. These certificates are typically used in non-production cases, like development and testing. One way to provision these kinds of certificates is through infrastructure as code. This allows consistent and repeatable certificate provisioning without the overhead of involving established CAs like Vault for signing and verifying. 
```

<ImageConfig hideBorder caption="Provision TLS certificates with Terraform and the TLS provider">

![Provision TLS certificates with Terraform and the TLS provider](/img/well-architected/diagram-terraform-tls-provider.png#light-theme-only)

</ImageConfig>

<ImageConfig hideBorder>

![Provision TLS certificates with Terraform and the TLS provider](/img/well-architected/diagram-terraform-tls-provider-dark.png#dark-theme-only)

</ImageConfig>

> ### Provision TLS certificates and keys

Reviewer: ```suggestion
## Provision TLS certificates and keys
```

> Terraform providers can be thought of as clients or plugins for an API, and they serve as the actor that transforms configuration resources into infrastructure. You can use infrastructure as code to create and manage the lifecycle of self-signed certificates with Terraform and the TLS provider. The TLS provider has utilities for working with Transport Layer Security keys and certificates. It also provides resources for creating private keys, certificates, and certificate requests as part of a Terraform deployment. Some benefits from using the Terraform TLS provider to manage your self-signed certificates include:

Reviewer: ```suggestion
Terraform providers are similar to API clients or plugins, serving as the actor that transforms configuration resources into infrastructure. You can use infrastructure as code to create and manage the lifecycle of self-signed certificates with Terraform and the TLS provider. The TLS provider lets you interact with TLS keys and certificates by providing resources for private keys, certificates, and certificate requests. Some benefits from using the Terraform TLS provider to manage your self-signed certificates include:
```

> - The TLS provider enables you to manage your certificates and keys with infrastructure as code.
> - You can use the TLS provider to automate your certificate and key issuance and lifecycle management.

Reviewer: ```suggestion
- The TLS provider lets you to manage your certificates and keys with infrastructure as code, enabling you to automate your certificate and key issuance and lifecycle.
```

> - Your TLS configuration is reproducible across different environments, and you can avoid configuration drift and related issues.

Reviewer: ```suggestion
- Your TLS configuration is reproducible across different environments and you can avoid configuration drift and related issues.
```

- By managing your TLS configuration with Terraform, you gain an audit trail of all changes to your configuration.

HashiCorp resources:

- [Terraform TLS provider](https://registry.terraform.io/providers/hashicorp/tls/latest/docs)

> ### Use Automated Certificate Management Environment (ACME)

Reviewer: ```suggestion
## Use Automated Certificate Management Environment (ACME)
```

ACME simplifies certificate lifecycle management by providing a standardized protocol for the automation of common operations, such as certificate requests and renewals. If you use an external CA like Let's Encrypt, you can use the ACME provider. ACME is an evolving standard for automating a domain-validated certificate authority. Clients register on an authority using a private key and contact information and answer challenges for claimed domains by supplying data issued by the authority through HTTP or DNS. Through this process, clients prove that they own the domains in question and can request certificates via the CA. No part of this process requires user interaction, which is a conventional blocker in obtaining a domain-validated certificate.

HashiCorp resources:

- [ACME Certificate and Account Provider](https://registry.terraform.io/providers/vancluever/acme/latest/docs)
- [Enable ACME with PKI secrets engine](/vault/tutorials/secrets-management/pki-acme-caddy)

## Next steps

> In this guide you learned how to use HashiCorp tools to secure digital entities with X.509 certificates, along with the benefits you gain when you do so. Specifically, you learned how to use Vault as a certificate authority and to secure application workloads with TLS authentication. You also learned how to use Consul to secure application workloads, and how Terraform can assist you with provisioning self-signed certificates and using ACME. You can continue to learn more about [best practices to protect sensitive data](/well-architected-framework/security/security-sensitive-data) in our Well-Architected Framework content.

Reviewer: ```suggestion
In this document, you learned how to use HashiCorp tools to secure digital entities with X.509 certificates, along with the benefits you gain when you do so. Specifically, you learned how to use Vault as a certificate authority and to secure application workloads with TLS authentication. You also learned how to use Consul to secure application workloads, and how Terraform can assist you with provisioning self-signed certificates and using ACME. You can continue to learn more about [securing your systems](/well-architected-framework/secure-systems).
```
